src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/MapCode/AbstractMapCodeService.cs

```diff
@@ -0,0 +1,356 @@
+        {
+            // We only support document edits for now. In the future once the client supports it, we should look
+            // into also supporting file creation/deletion/rename.
+            foreach (var edit in documentEdits)
+            {
+                var success = await TryProcessEditAsync(documentContext, edit.TextDocument.DocumentUri.GetRequiredParsedUri(), edit.Edits, csharpChanges, cancellationToken).ConfigureAwait(false);
+                if (!success)
+                {
+                    return false;
+                }
+            }
+        }
+
+        if (csharpEdit.Changes is not null)
+        {
+            foreach (var edit in csharpEdit.Changes)
+            {
+                var generatedUri = new Uri(edit.Key);
+                var success = await TryProcessEditAsync(documentContext, generatedUri, [.. edit.Value.Select(e => (SumType<TextEdit, AnnotatedTextEdit>)e)], csharpChanges, cancellationToken).ConfigureAwait(false);
+                if (!success)
+                {
+                    return false;
+                }
+            }
+        }
+
+        changes.AddRange(csharpChanges);
+        return true;
+
+        async Task<bool> TryProcessEditAsync(
+            DocumentContext documentContext,
+            Uri generatedUri,
+            SumType<TextEdit, AnnotatedTextEdit>[] textEdits,
+            List<TextDocumentEdit> csharpChanges,
+            CancellationToken cancellationToken)
+        {
+            foreach (TextEdit documentEdit in textEdits)
+            {
+                var (hostDocumentUri, hostDocumentRange) = await MapToHostDocumentUriAndRangeAsync(documentContext, generatedUri, documentEdit.Range.ToLinePositionSpan(), cancellationToken).ConfigureAwait(false);
+
+                if (hostDocumentUri == generatedUri)
+                {
+                    return false;
+                }
+
+                var textEdit = LspFactory.CreateTextEdit(hostDocumentRange, documentEdit.NewText);
+
+                var textDocumentEdit = new TextDocumentEdit
+                {
+                    TextDocument = new OptionalVersionedTextDocumentIdentifier { DocumentUri = new(hostDocumentUri) },
+                    Edits = [textEdit]
+                };
+
+                csharpChanges.Add(textDocumentEdit);
+            }
+
+            return true;
+        }
+    }
+
+    // Resolve edits that are at the same start location by merging them together.
+    public static void MergeEdits(List<TextDocumentEdit> changes)
+    {
+        var groupedChanges = changes.GroupBy(c => c.TextDocument.DocumentUri).ToImmutableArray();
+        changes.Clear();
+        foreach (var documentChanges in groupedChanges)
+        {
+            var edits = documentChanges.ToList();
+            edits.Sort((x, y) => ((TextEdit)x.Edits.Single()).Range.Start.CompareTo(((TextEdit)y.Edits.Single()).Range.Start));
```

Comment:
Isn't `TextDocumentEdit.Edits` just an array? If so, why not use an indexer (`[0]`) rather than all of the `Single()` calls? If it's necessary to validate that each `Edits` array contains one and only one item, it'd be better to do that up front in a single loop rather than in every comparison made during a sort.